File: app/src/main/java/com/minkiapps/livetranslator/TTSWrapper.java
package com.minkiapps.livetranslator;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Pair;

import com.huawei.hms.mlsdk.tts.MLTtsAudioFragment;
import com.huawei.hms.mlsdk.tts.MLTtsCallback;
import com.huawei.hms.mlsdk.tts.MLTtsConfig;
import com.huawei.hms.mlsdk.tts.MLTtsConstants;
import com.huawei.hms.mlsdk.tts.MLTtsEngine;
import com.huawei.hms.mlsdk.tts.MLTtsError;
import com.huawei.hms.mlsdk.tts.MLTtsWarn;

import timber.log.Timber;

public class TTSWrapper { //TODO need to find out why this has to be inside a java class

    private MLTtsEngine mlCNTtsEngine;
    private MLTtsEngine mlENTtsEngine;
    private TTSListener ttsListener;

    private final Handler handler = new Handler(Looper.getMainLooper());

    public TTSWrapper() {
        createCNTtsEngine();
        createENTtsEngine();
    }

    public void speakEnglish(final String text) {
        mlENTtsEngine.speak(text, MLTtsEngine.QUEUE_FLUSH);
    }

    public void speakChinese(final String text) {
        mlCNTtsEngine.speak(text, MLTtsEngine.QUEUE_FLUSH);
    }

    private void createCNTtsEngine() {
        MLTtsConfig mlConfigs = new MLTtsConfig()
                .setLanguage(MLTtsConstants.TTS_ZH_HANS)
                .setPerson(MLTtsConstants.TTS_SPEAKER_FEMALE_ZH)
                .setSpeed(1f)
                .setVolume(1f);
        MLTtsCallback callback = new MLTtsCallback() {
            @Override
            public void onError(String taskId, MLTtsError err) {
                if(ttsListener != null) {
                    handler.post(() -> ttsListener.onError(err));
                }
            }

            @Override
            public void onWarn(String taskId, MLTtsWarn warn) {
                Timber.w("TaskId: " + taskId + " Message " + warn.getWarnMsg());
            }

            @Override
            public void onRangeStart(String taskId, int start, int end) {
                Timber.d("TaskId: " + taskId + " Start: " + start + " End: " + end);
            }

            @Override
            public void onAudioAvailable(String s, MLTtsAudioFragment mlTtsAudioFragment, int i, Pair<Integer, Integer> pair, Bundle bundle) {
                Timber.d(s);
            }

            @Override
            public void onEvent(String taskId, int eventName, Bundle bundle) {
                switch (eventName) {
                    case MLTtsConstants.EVENT_PLAY_STOP:
                        if(ttsListener != null) {
                            handler.post(() -> ttsListener.onSpeechEnded());
                        }
                        break;
                }
            }
        };
        mlCNTtsEngine = new MLTtsEngine(mlConfigs);
        mlCNTtsEngine.setTtsCallback(callback);
    }

    private void createENTtsEngine() {
        MLTtsConfig mlConfigs = new MLTtsConfig()
                .setLanguage(MLTtsConstants.TTS_EN_US)
                .setPerson(MLTtsConstants.TTS_SPEAKER_FEMALE_EN)
                .setSpeed(1f)
                .setVolume(1f);
        MLTtsCallback callback = new MLTtsCallback() {
            @Override
            public void onError(String taskId, MLTtsError err) {
                if(ttsListener != null) {
                    handler.post(() -> ttsListener.onError(err));
                }
            }

            @Override
            public void onWarn(String taskId, MLTtsWarn warn) {
                Timber.w("TaskId: " + taskId + " Message " + warn.getWarnMsg());
            }

            @Override
            public void onRangeStart(String taskId, int start, int end) {
                Timber.d("TaskId: " + taskId + " Start: " + start + " End: " + end);
            }

            @Override
            public void onAudioAvailable(String s, MLTtsAudioFragment mlTtsAudioFragment, int i, Pair<Integer, Integer> pair, Bundle bundle) {
                Timber.d(s);
            }

            @Override
            public void onEvent(String taskId, int eventName, Bundle bundle) {
                switch (eventName) {
                    case MLTtsConstants.EVENT_PLAY_STOP:
                        if(ttsListener != null) {
                            handler.post(() -> ttsListener.onSpeechEnded());
                        }
                        break;
                }
            }
        };
        mlENTtsEngine = new MLTtsEngine(mlConfigs);
        mlENTtsEngine.setTtsCallback(callback);
    }

    public void stop() {
        mlENTtsEngine.stop();
        mlCNTtsEngine.stop();
    }

    public void shutdown() {
        mlENTtsEngine.shutdown();
        mlCNTtsEngine.shutdown();
    }

    public void setTtsListener(TTSListener ttsListener) {
        this.ttsListener = ttsListener;
    }

    interface TTSListener {
        void onSpeechEnded();
        void onError(MLTtsError err);
    }
}

File: app/src/main/java/com/minkiapps/livetranslator/TTSWrapper.kt
package com.minkiapps.livetranslator

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Pair
import com.huawei.hms.mlsdk.tts.*
import com.minkiapps.livetranslator.translation.Translation
import timber.log.Timber
import kotlin.collections.HashMap

class TTSWrapper {

    private lateinit var mlCNTtsEngine: MLTtsEngine
    private lateinit  var mlENTtsEngine: MLTtsEngine
    private var ttsListener: TTSListener? = null

    private val engineMap : HashMap<String, MLTtsEngine> = HashMap()

    private val handler = Handler(Looper.getMainLooper())

    @Synchronized
    fun speak(translation: Translation, text : String) {

    }

    fun speakEnglish(text: String) {
        mlENTtsEngine.speak(text, MLTtsEngine.QUEUE_FLUSH)
    }

    fun speakChinese(text: String) {
        mlCNTtsEngine.speak(text, MLTtsEngine.QUEUE_FLUSH)
    }

    private fun createCNTtsEngine() {
        val mlConfigs = MLTtsConfig()
            .setLanguage(MLTtsConstants.TTS_ZH_HANS)
            .setPerson(MLTtsConstants.TTS_SPEAKER_FEMALE_ZH)
            .setSpeed(1f)
            .setVolume(1f)
        val callback: MLTtsCallback = object : MLTtsCallback {
            override fun onError(taskId: String, err: MLTtsError) {
                if (ttsListener != null) {
                    handler.post { ttsListener?.onError(err) }
                }
            }

            override fun onWarn(taskId: String, warn: MLTtsWarn) {
                Timber.w("TaskId: $taskId Message ${warn.warnMsg}")
            }

            override fun onRangeStart(taskId: String, start: Int, end: Int) {
                Timber.w("TaskId: $taskId Start $start End: $end")
            }

            override fun onAudioAvailable(
                s: String,
                mlTtsAudioFragment: MLTtsAudioFragment,
                i: Int,
                pair: Pair<Int, Int>,
                bundle: Bundle
            ) {
                Timber.d(s)
            }

            override fun onEvent(taskId: String, eventName: Int, bundle: Bundle?) {
                when (eventName) {
                    MLTtsConstants.EVENT_PLAY_STOP -> if (ttsListener != null) {
                        handler.post { ttsListener?.onSpeechEnded() }
                    }
                }
            }
        }
        mlCNTtsEngine = MLTtsEngine(mlConfigs)
        mlCNTtsEngine.setTtsCallback(callback)
    }

    private fun createENTtsEngine() {
        val mlConfigs = MLTtsConfig()
            .setLanguage(MLTtsConstants.TTS_EN_US)
            .setPerson(MLTtsConstants.TTS_SPEAKER_FEMALE_EN)
            .setSpeed(1f)
            .setVolume(1f)
        val callback: MLTtsCallback = object : MLTtsCallback {
            override fun onError(taskId: String, err: MLTtsError) {
                if (ttsListener != null) {
                    handler.post { ttsListener?.onError(err) }
                }
            }

            override fun onWarn(taskId: String, warn: MLTtsWarn) {
                Timber.w("TaskId: $taskId Message ${warn.warnMsg}")
            }

            override fun onRangeStart(taskId: String, start: Int, end: Int) {
                Timber.d("TaskId: $taskId Start: $start End: $end")
            }

            override fun onAudioAvailable(
                s: String,
                mlTtsAudioFragment: MLTtsAudioFragment,
                i: Int,
                pair: Pair<Int, Int>,
                bundle: Bundle
            ) {
                Timber.d(s)
            }

            override fun onEvent(taskId: String, eventName: Int, bundle: Bundle?) {
                when (eventName) {
                    MLTtsConstants.EVENT_PLAY_STOP -> if (ttsListener != null) {
                        handler.post { ttsListener?.onSpeechEnded() }
                    }
                }
            }
        }
        mlENTtsEngine = MLTtsEngine(mlConfigs)
        mlENTtsEngine.setTtsCallback(callback)
    }

    fun stop() {
        //mlENTtsEngine.stop()
        //mlCNTtsEngine.stop()
    }

    fun shutdown() {
        //mlENTtsEngine.shutdown()
        //mlCNTtsEngine.shutdown()
    }

    fun setTtsListener(ttsListener: TTSListener) {
        this.ttsListener = ttsListener
    }

    interface TTSListener {
        fun onSpeechEnded()
        fun onError(err: MLTtsError)
    }
}